<?xml version="1.0"?>
<sdf version="1.4">
    <model name="rover">
        <static>false</static>
        <pose>0 0 .15 0 0 0</pose>

        <link name="camera_link">
            <pose>0.3 0 0.27 0 0 0</pose>
            <inertial>
                <mass>0.15</mass>
            </inertial>
            <collision name="camera_collision">
                <geometry>
                    <box>
                        <size>0.05 0.1 0.1</size>
                    </box>
                </geometry>
            </collision>
            <visual name="camera_visual">
                <geometry>
                    <box>
                        <size>0.05 0.1 0.1</size>
                    </box>
                </geometry>
            </visual>
            <sensor name="camera" type="camera">
                <camera>
                    <horizontal_fov>1.045</horizontal_fov>
                    <image>
                        <width>320</width>
                        <height>240</height>
                    </image>
                    <clip>
                        <near>0.1</near>
                        <far>100</far>
                    </clip>
                </camera>
                <always_on>1</always_on>
                <update_rate>25</update_rate>
                <visualize>true</visualize>
            </sensor>
        </link>

        <joint name="camera_joint" type="revolute">
            <parent>chassis</parent>
            <child>camera_link</child>
            <axis>
                <xyz>0 0 1</xyz>
                <limit>
                    <upper>0</upper>
                    <lower>0</lower>
                </limit>
            </axis>
        </joint>

        <link name="chassis">
            <pose>-0.02 0 0.15 0 0 0</pose>
            <inertial>
                <mass>10</mass>
            </inertial>
            <collision name="chassis_collision">
                <geometry>
                    <box>
                        <size>.8 .4 .16</size>
                    </box>
                </geometry>
            </collision>
            <visual name="chassis_visual">
                <geometry>
                    <box>
                        <size>.8 .4 .16</size>
                    </box>
                </geometry>
            </visual>
        </link>


        <link name="rear_left_wheel">
            <pose>-0.3 0.24 0.1 0 1.5707 1.5707</pose>
            <inertial>
                <mass>1.5</mass>
            </inertial>
            <collision name="rear_left_wheel_collision">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
                <surface>
                    <friction>
                        <ode>
                            <mu>100</mu>
                            <mu2>50</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <visual name="rear_left_wheel_visual">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
            </visual>
        </link>

        <joint type="revolute" name="rear_left_joint">
            <child>rear_left_wheel</child>
            <parent>chassis</parent>
            <axis>
                <xyz>0 1 0</xyz>
            </axis>
        </joint>


        <link name="rear_right_wheel">
            <pose>-0.3 -0.24 0.1 0 1.5707 1.5707</pose>
            <inertial>
                <mass>1.5</mass>
            </inertial>
            <collision name="rear_right_wheel_collision">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
                <surface>
                    <friction>
                        <ode>
                            <mu>100</mu>
                            <mu2>50</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <visual name="rear_right_wheel_visual">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
            </visual>
        </link>

        <joint type="revolute" name="rear_right_joint">
            <child>rear_right_wheel</child>
            <parent>chassis</parent>
            <axis>
                <xyz>0 1 0</xyz>
            </axis>
        </joint>


        <link name="front_left_wheel">
            <pose>0.3 0.24 0.1 0 1.5707 1.5707</pose>
            <inertial>
                <mass>1.5</mass>
            </inertial>
            <collision name="front_left_wheel_collision">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
                <surface>
                    <friction>
                        <ode>
                            <mu>100</mu>
                            <mu2>50</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <visual name="front_left_wheel_visual">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
            </visual>
        </link>

        <link name="front_left_connector">
            <pose>0.3 0.25 0.1 0 1.5707 1.5707</pose>
        </link>

        <joint type="revolute" name="front_left_joint">
            <parent>chassis</parent>
            <child>front_left_connector</child>
            <axis>
                <xyz>0 0 1</xyz>
            </axis>
        </joint>

        <joint type="revolute" name="front_left_joint2">
            <parent>front_left_connector</parent>
            <child>front_left_wheel</child>
            <axis>
                <xyz>0 1 0</xyz>
            </axis>
        </joint>


        <link name="front_right_wheel">
            <pose>0.3 -0.24 0.1 0 1.5707 1.5707</pose>
            <inertial>
                <mass>1.5</mass>
            </inertial>
            <collision name="front_right_wheel_collision">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
                <surface>
                    <friction>
                        <ode>
                            <mu>100</mu>
                            <mu2>50</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <visual name="front_right_wheel_visual">
                <geometry>
                    <cylinder>
                        <radius>.2</radius>
                        <length>.07</length>
                    </cylinder>
                </geometry>
            </visual>
        </link>

        <link name="front_right_connector">
            <pose>0.3 -0.25 0.1 0 1.5707 1.5707</pose>
        </link>

        <joint type="revolute" name="front_right_joint">
            <parent>chassis</parent>
            <child>front_right_connector</child>
            <axis>
                <xyz>0 0 1</xyz>
                <limit>
                    <lower>-0.61</lower>
                    <upper>0.61</upper>
                </limit>
            </axis>
        </joint>

        <joint type="revolute" name="front_right_joint2">
            <parent>front_right_connector</parent>
            <child>front_right_wheel</child>
            <axis>
                <xyz>0 1 0</xyz>
            </axis>
        </joint>

        <plugin name="RoverPlugin" filename="libRoverPlugin.so">
            <chassis_name>chassis</chassis_name>
            <front_left_joint>front_left_joint</front_left_joint>
            <front_right_joint>front_right_joint</front_right_joint>
            <rear_left_joint>rear_left_joint</rear_left_joint>
            <rear_right_joint>rear_right_joint</rear_right_joint>
        </plugin>

    </model>
</sdf>

